src/runtime/device_interface.cpp

```diff
@@ -237,6 +237,10 @@ WEAK int halide_device_free(void *user_context, struct buffer_t *buf) {
     debug(user_context) << "halide_device_free: " << buf
                         << " buf dev " << buf->dev
                         << " interface " << halide_get_device_interface(dev_field) << "\n";
+    if (buf->dev_dirty) {
```

Comment:
Ok, i need to understand the use of halide_device_free better, I think because the question in my head is if a buffer is just an intermediate between two pipelines, then why would device_free be called on it?